plugins/ui/src/js/src/widget/WidgetUtils.test.tsx

```diff
@@ -246,3 +247,94 @@ describe('wrapCallable', () => {
     ).rejects.toBeInstanceOf(Error);
   });
 });
+
+describe('decodeNode', () => {
+  it.each([null, undefined, '', 'test', 0, true])(
+    'should handle primitive values: %s',
+    value => {
+      const callback = jest.fn((k, v) => v);
+      expect(decodeNode(value, callback)).toBe(value);
+      expect(callback).toHaveBeenCalledWith('', value);
+    }
+  );
+
+  describe('arrays', () => {
+    it('returns original array if no children changed', () => {
+      const callback = jest.fn((k, v) => v);
+      const value = ['test', 'test2'];
+      expect(decodeNode(value, callback)).toBe(value);
+      expect(callback).toHaveBeenCalledTimes(3);
+      expect(callback).toHaveBeenCalledWith('0', 'test');
+      expect(callback).toHaveBeenCalledWith('1', 'test2');
+      expect(callback).toHaveBeenCalledWith('', value);
+    });
+
+    it('returns a new array if children changed', () => {
+      const callback = jest.fn((k, v) => (k === '0' ? 'modified' : v));
+      const value = ['test', 'test2'];
+      const expected = ['modified', 'test2'];
+      const result = decodeNode(value, callback);
+      expect(result).toEqual(expected);
+      expect(result).not.toBe(value);
+    });
+
+    it('handles the empty array', () => {
+      const callback = jest.fn((k, v) => v);
+      const value: unknown[] = [];
+      const result = decodeNode(value, callback);
+      expect(result).toEqual([]);
+      expect(result).toBe(value);
+    });
+  });
+
+  describe('objects', () => {
+    it('returns original object if no children changed', () => {
+      const callback = jest.fn((k, v) => v);
+      const value = { a: 'test', b: 'test2' };
+      expect(decodeNode(value, callback)).toBe(value);
```

Comment:
Same as array suggestion

```suggestion
      const value = { a: 'test', b: 'test2' };
      const valueCopy = { ...value };
      const result = decodeNode(value, callback);
      expect(result).toBe(value);
      expect(result).toStrictEqual(valueCopy);
```